### PR TITLE
Fixes #15926 - [API] Add method to get ENC values of a host

### PR DESCRIPTION
**Pull request by soukron**
Fixes #15926 - [API] Add method to get ENC values of a host

### CODE REVIEW

**Comment by tbrisker** [comment]
[test] - this will trigger automatic testing on this PR.

**Comment by dLobatog** [comment]
@soukron Thanks for the patch!  

However, I don't follow what exactly you need this endpoint for? The ENC output is accessible at  /hosts/<fqdn>/externalNodes as the bug in redmine mentions, so I don't see what you want to accomplish by putting this in API too?

The ticket mentions ' I'm trying to get all parameters' for a host - those are served at `api/v2/hosts/$yourhostfqdn` , via the key `:all_parameters`.

**Comment by soukron** [comment]
Hi @dLobatog 
I'm using Satellite 6.1.9 which doesn't have such key `:all_parameters`. Also, Satellite 6.2 do has it but it still doesn't have Puppet classes parameters so it's not exactly the same than `/hosts/:id/externalNodes` information.

I've hacked also that URL to accept `format=json` but it still uses a different authentication method than API so it's not possible to use it as REST endpoint. That being said, I still see a valid point to have an API endpoint which returns the ENC information of a host. 

I also anticipate a use case where a user can get all this ENC stuff and save it in a hiera directory and export it to proxies running as Puppet Masters so a proxy can be disconnected from Foreman server even for Puppet runs. But this last point is another story :smile: One step at time.

Please let me know if you still have any question.

**Comment by soukron** [comment]
Hi again @dLobatog.
**TL;DR** the fact that a UI URL (made for browsers with cookie-based authentication) provides the information doesn't mean that having an API endpoint (made for other applications with basic/token authentication) is duplicating functions. If that were true, there won't be API endpoints at all as everything is already in UI URLs. Isn't it?

Let me try to be more specific.

See a regular API call:

```
$ curl --basic --user admin:admin123          \
       -H 'Accept: application/json;version=2' \
       https://sat6.example.com/api/v2/hosts/client1.example.com

{
 "name": "client1.example.com",
 "id": 27,
 "ip": "10.250.178.14",
 "environment_id": 141,
...
}
```

Now, an API call to this new endpoint:

```
$ curl --basic --user admin:admin123          \
       -H 'Accept: application/json;version=2' \
       https://sat6.example.com/api/v2/hosts/client1.example.com/enc

{
  "data": {
    "classes": {
      "db_class": {
        "mysql_fs_apps_size": "256M",
        "mysql_fs_data_size": "2048M",
      },
      "java_class": {
        "jre_fs_java_app_size": "256M",
        "jre_fs_java_log_size": "512M",
        "jre_java_password": "$6$BgbR4T9o$lE6NIfWt4DQ4lRDcwcJC5nztxohBbV5Wkmpw9RZj5ESYjRp3UfT2EVa527
      },
      "custom_stages": null,
    }
... rest of classes classes and their parameters already calculated ...
    "parameters": {
      "puppetmaster": "cap1.example.com",
      "domainname": "",
      "hostgroup": "qcc2016-20.93-caj-64/itg",
      "location": "Location",
      "organization": "Organization",
      "root_pw": "$6$36da.fmrOkMh4e0SVoR.hnyK8E58/CeKtSab6IhMISkjxU/nS4y9bKySxyrfgzVIdUFwlp05H6VnPCZ5kKmEy1YghNGkNjSnJLAH1",
    },
    "environment": "KT_Organization_itg_qcc2016_20_93_caj_64_87"
  }
}
```

Finally, let's see how it goes with externalNodes URL:

```
$ curl --basic --user admin:admin123          \
       -H 'Accept: application/json;version=2' \
       https://sat6.example.com/hosts/client1.example.com/externalNodes
[empty answer]
```

To get information from externalNodes URL I must login before (send a valid form to /users/login and retrieve the cookie) and send it later:

```
$ curl -b cookies.txt \
       https://sat6.example.com/hosts/client1.example.com/externalNodes

<pre>---
classes:
  db_bal:
    mysql_fs_apps_size: 256M
    mysql_fs_data_size: 2048M
...
  kt_org: Organization
environment: KT_Organization_itg_qcc2016_20_93_caj_64_87
```

As I mention, a URL is not an API endpoint. API endpoints use HTTP authentication while UI URLs use cookie-based authentication. It may be the same for a regular user but this is definitively a stopper when you're trying to create an external application which interacts with Foreman using API. Not to mention naming conventions for API endpoints.

**Comment by dLobatog** [comment]
@soukron I agree the authentication methods are different, my point was mostly that if the whole issue is getting the host parameters, you can get that at `:all_parameters`. 

However I'm looking further into this issue and I saw you want to extract the ENC info to another system which is indeed a bit more tricky & the API endpoint makes sense for that use case. I'll review & test this patch - keep in mind you can use `/etc/puppet/node.rb` to get the ENC for any host easily already, in case you need a workaround for the moment.

Thanks again!

**Comment by dLobatog** [comment]
@soukron I agree the authentication methods are different, my point was mostly that if the whole issue is getting the host parameters, you can get that at `:all_parameters`. 

However I'm looking further into this issue and I saw you want to extract the ENC info to another system which is indeed a bit more tricky & the API endpoint makes sense for that use case. I'll review & test this patch - keep in mind you can use `/etc/puppet/node.rb` to get the ENC for any host easily already, in case you need a workaround for the moment.

Thanks again!

**Comment by dLobatog** [comment]
@soukron I agree the authentication methods are different, my point was mostly that if the whole issue is getting the host parameters, you can get that at `:all_parameters`. 

However I'm looking further into this issue and I saw you want to extract the ENC info to another system which is indeed a bit more tricky & the API endpoint makes sense for that use case. I'll review & test this patch - keep in mind you can use `/etc/puppet/node.rb` to get the ENC for any host easily already, in case you need a workaround for the moment.

Thanks again!

edit - this message was sent 5 times, sorry

**Comment by dLobatog** [comment]
@soukron I agree the authentication methods are different, my point was mostly that if the whole issue is getting the host parameters, you can get that at `:all_parameters`. 

However I'm looking further into this issue and I saw you want to extract the ENC info to another system which is indeed a bit more tricky & the API endpoint makes sense for that use case. I'll review & test this patch - keep in mind you can use `/etc/puppet/node.rb` to get the ENC for any host easily already, in case you need a workaround for the moment.

Thanks again!

**Comment by dLobatog** [comment]
@soukron I agree the authentication methods are different, my point was mostly that if the whole issue is getting the host parameters, you can get that at `:all_parameters`. 

However I'm looking further into this issue and I saw you want to extract the ENC info to another system which is indeed a bit more tricky & the API endpoint makes sense for that use case. I'll review & test this patch - keep in mind you can use `/etc/puppet/node.rb` to get the ENC for any host easily already, in case you need a workaround for the moment.

Thanks again!

File: app/controllers/api/v2/hosts_controller.rb [hunk 1]
```diff
@@ -126,6 +126,13 @@ def destroy
         process_response @host.destroy
       end
 
+      api :GET, "/hosts/:id/enc", N_("Get ENC values of host")
+      param :id, :identifier_dottable, :required => true
+
+      def enc
+        render :json => { :data => @host.info }.to_json if @host
```
Comment (dLobatog) [inline app/controllers/api/v2/hosts_controller.rb]: Isn't the `.to_json` here redundant? `render :json =>` will already transform the hash to JSON I think

Comment (dLobatog) [inline app/controllers/api/v2/hosts_controller.rb]: Also the `if @host` is redundant I think, you already specify the parameter as `:required => true` above. If you look at the top of the file, the `before_action :find_resource` will take care of finding `@host` and failing with 404 if it can't be found, or 403 or whatever.

**Comment by dLobatog** [comment]
@soukron Thanks! It's all fine in my opinion, I left a couple of comments inline about two things that I think are unnecessary in the controller.

Can you write a couple of tests for this? See `test/functional/api/v2/hosts_controller.rb` for examples. Just verifying it returns the host.info in the normal use case should be fine. Please don't stub `host.info`  in this case.

You can run the tests with `ruby -I'test' test/functional/api/v2/hosts_controller_test.rb`. It should be ready to merge after that. Thanks again!

File: app/controllers/api/v2/hosts_controller.rb (hunk 1 above)
Comment (soukron) [inline app/controllers/api/v2/hosts_controller.rb]: I also thought it could be redundant but as I saw the same checks in `status/` call I did it in the same way.

See line 151: it checks `if @host` and also applies `.to_json`.

**Comment by soukron** [comment]
> keep in mind you can use `/etc/puppet/node.rb` to get the ENC for any host easily already

Sure! Actually, `node.rb` was my first attempt to accomplish the task but: 
- `node.rb` uses files stored in `/var/lib/puppet/yaml/` and for some reason (which is unknown for me, yet) I have all hosts' files distributed among Satellite and Capsules so I hadn't a single point to get information of all hosts.
- I realized that the same code used in externalNodes could work in a new method in host controller.

I'll do the tests and push the changes.

**Comment by soukron** [comment]
I apologize for the delay but I haven't found time yet to write down the tests.

**Comment by tbrisker** [comment]
Please disregard the messages from the hound, once you rebase your PR on develop they will go away (because your branch lacks a hound configuration, hound uses it's defaults which are different then what we use)

**Comment by dLobatog** [comment]
@soukron It seems that there are still conflicts, please rebase the branch (`git pull --rebase upstream develop` for a upstream remote that points to theforeman/foreman) - I'll take a look in a min :smile:

**Comment by ares** [comment]
@xprazak2 kindly rebased and opened as a separate PR at https://github.com/theforeman/foreman/pull/3907

**Comment by dLobatog** [comment]
Closed because #3907 was merged instead
